dali/kernels/slice/slice_gpu.cuh

```diff
@@ -261,11 +261,15 @@ class SliceGPU {
       total_volume += volume(args.shape);
     }
 
-    unsigned min_blocks = 4 * GetSmCount();
-    block_size_ = kMaxBlockSize;
-    while (total_volume / block_size_ < min_blocks && block_size_ > kMinBlockSize) {
-      block_size_ /= 2;
-    }
+    int blocksPerSM;
+    CUDA_CALL(cudaOccupancyMaxActiveBlocksPerMultiprocessor(&blocksPerSM,
+              detail::SliceKernel<OutputType, InputType, Dims, false>, kBlockDim, 0));
+    unsigned blocks_in_wave = blocksPerSM * GetSmCount();
+    unsigned waves = div_ceil(total_volume, kMaxBlockSize * blocks_in_wave);
+    unsigned values_per_thread = div_ceil(total_volume, blocks_in_wave * kBlockDim * waves);
+    block_size_ = values_per_thread * kBlockDim;
```

Comment:
How about
```suggestion
    block_size_  = div_ceil(total_volume, blocks_in_wave * waves);
```

Reply:
Or ceiling is the key here?

Reply:
This ceiling is there to ensure that `block_size_` (which is count of values to be processed by a block) is a multiply of `kBlockDim = 256`, i.e. that each thread in a block will process same count of values (`values_per_thread`).

Reply:
The ceiling is there to ensure that `block_size_` (number of values to be processed by a block) is a multiply of `kBlockDim = 256`, i.e. that each thread will do the same amount of work.

Reply:
Uneven amount of work within a block is fine as long as it's the same for each warp.
Also, please use CUDA nomenclature (warp, not wave).

Reply:
@mzient thank you! Seems that my intuition was wrong. I'll benchmark the code with @JanuszL's simplification to make sure it's still fine and if it is, I'll apply the change.

Reply:
You were right – no performance change at all. Applied in https://github.com/NVIDIA/DALI/pull/3604/commits/193690151747a5a2ad92bdddbc9f6ee392978a9e